### PR TITLE
Add a function to activate CW mode

### PR DESCRIPTION
**Pull request by Miceuz**
It's just a wrapper function for underlying Semtech infrastructure for activating CW mode.

### CODE REVIEW

**Review by matthijskooijman — CHANGES_REQUESTED** [review]
Thanks!

Overall, this PR looks good. Some remarks/questions:
 - You've also included a commit for RAK support, that probably warrants its own PR. Could you remove it (just force push to your branch with the commit removed).
 - Are you sure the power is directly specified in dB? The regular TX power is specified by index, with a non-trivial conversion from dB: https://github.com/stm32duino/STM32LoRaWAN/blob/46d7e916f879bf5f5e463b1f71e216dc6b4d04e3/src/STM32LoRaWAN.cpp#L218-L223
 - Is there a special "no timeout" value? If so, the comments should specify it.

**Comment by Miceuz** [comment]
Argh, you are right, I have based my changes on previous branch. Will fix that.

You are right about power units also.

**Comment by Miceuz** [comment]
Fixed that

**Comment by matthijskooijman** [comment]
Seems you dropped the RAK commit, but haven't addressed the other questions about power and timeout yet?

**Comment by Miceuz** [comment]
> Seems you dropped the RAK commit, but haven't addressed the other questions about power and timeout yet?

Yes, I have removed dBm mention from the documentation and no, there is no special "no timeout" value it seems:

```c
UTIL_TIMER_Status_t UTIL_TIMER_SetPeriod(UTIL_TIMER_Object_t *TimerObject, uint32_t NewPeriodValue)
{
  UTIL_TIMER_Status_t  ret = UTIL_TIMER_OK;
  
  if(NULL == TimerObject)
  {
	  ret = UTIL_TIMER_INVALID_PARAM;
  }
  else
  {
    TimerObject->ReloadValue = UTIL_TimerDriver.ms2Tick(NewPeriodValue);
    if(TimerExists(TimerObject))
    {
      (void)UTIL_TIMER_Stop(TimerObject);
      ret = UTIL_TIMER_Start(TimerObject);
    }
  }
  return ret;
}
```

Time is converted from seconds to ms in `radio.c`:
```c
    uint32_t timeout = ( uint32_t )time * 1000;
    uint8_t antswitchpow;

    SUBGRF_SetRfFrequency( freq );

    antswitchpow = SUBGRF_SetRfTxPower( power );

    /* WORKAROUND - Trimming the output voltage power_ldo to 3.3V */
    SUBGRF_WriteRegister(REG_DRV_CTRL, 0x7 << 1);

    /* Set RF switch */
    SUBGRF_SetSwitch( antswitchpow, RFSWITCH_TX );

    SUBGRF_SetTxContinuousWave( );

    TimerSetValue( &TxTimeoutTimer, timeout );
    TimerStart( &TxTimeoutTimer );
```

File: src/STM32LoRaWAN.h [hunk 1]
```diff
@@ -791,6 +791,19 @@ class STM32LoRaWAN : public Stream {
 
     /// @}
 
+    /**
+     * Set the radion into continuous wave (CW) mode. In this mode radio outputs
```
Comment (fpistm) [inline src/STM32LoRaWAN.h:795]:
```suggestion
     * Set the radio into continuous wave (CW) mode. In this mode radio outputs
```

File: src/STM32LoRaWAN.h [hunk 2]
```diff
@@ -791,6 +791,19 @@ class STM32LoRaWAN : public Stream {
...
+     * a signal at the specified frequency and power for the specified duration. 
+     * 
+     * Used only for compliance testing and antenna testing.
+     *
+     * \param frequency Frequency in Hz.
+     * \param power TX power level [0-15].
+     * \param timeout Duration in seconds.
+     * \returns True if CW mode was activated successfully.
```
Comment (fpistm) [inline src/STM32LoRaWAN.h:803]:
```suggestion
     * \returns true if CW mode was activated successfully.
```

**Review by fpistm — CHANGES_REQUESTED** [review]
Small typos else LGTM.
Thanks for your contribution.

**Comment by matthijskooijman** [comment]
> Yes, I have removed dBm mention from the documentation and no, there is no special "no timeout" value it seems:

I wonder if the API *should* accept a power in dBm, just like the regular power setting function. Of course it is a testing function, so having a somewhat less friendly API that has a potentially hardware-dependent power argument might be acceptable (@fpistm, what do you think?), but in that case at least the documentation should say *something* about what the power values mean (even if it is just a reference to the relevant register in the STM32WL55 datasheet).

**Comment by fpistm** [comment]
> > Yes, I have removed dBm mention from the documentation and no, there is no special "no timeout" value it seems:
> 
> I wonder if the API _should_ accept a power in dBm, just like the regular power setting function. Of course it is a testing function, so having a somewhat less friendly API that has a potentially hardware-dependent power argument might be acceptable (@fpistm, what do you think?), but in that case at least the documentation should say _something_ about what the power values mean (even if it is just a reference to the relevant register in the STM32WL55 datasheet).

Agreed.
`* \param power TX power level [0-15].` is confusing as it seems it should be in dBm.
So I would rewrite this: `* \param powerdBm transmit power in dBm.` and rename arg `powerdBm`

**Comment by Miceuz** [comment]
> > > Yes, I have removed dBm mention from the documentation and no, there is no special "no timeout" value it seems:
> > 
> > 
> > I wonder if the API _should_ accept a power in dBm, just like the regular power setting function. Of course it is a testing function, so having a somewhat less friendly API that has a potentially hardware-dependent power argument might be acceptable (@fpistm, what do you think?), but in that case at least the documentation should say _something_ about what the power values mean (even if it is just a reference to the relevant register in the STM32WL55 datasheet).
> 
> Agreed. `* \param power TX power level [0-15].` is confusing as it seems it should be in dBm. So I would rewrite this: `* \param powerdBm transmit power in dBm.` and rename arg `powerdBm`

Cool, I will re-calculate to index from dB in the same manner you have posted above `int8_t index = -(db - 1) / 2; `

**Comment by fpistm** [comment]
Looking at the code, when `LoRaMacMlmeRequest` is called, it calls:
https://github.com/stm32duino/STM32LoRaWAN/blob/d3aa3852d36afe0ad4e46d0a4524cc13957175fc/src/STM32CubeWL/LoRaWAN/Mac/LoRaMac.c#L6247
then `SetTxContinuousWave`:
https://github.com/stm32duino/STM32LoRaWAN/blob/d3aa3852d36afe0ad4e46d0a4524cc13957175fc/src/STM32CubeWL/LoRaWAN/Mac/LoRaMac.c#L4093
with `SetTxContinuousWave `:
https://github.com/stm32duino/STM32LoRaWAN/blob/d3aa3852d36afe0ad4e46d0a4524cc13957175fc/src/STM32CubeWL/SubGHz_Phy/radio.h#L297-L304

So I think we could only deal with dBm while it is explicitly describe, as this feature is for test purpose, I think it is acceptable.

 @matthijskooijman ?

**Comment by Miceuz** [comment]
Yes, and then `RadioSetTxContinuousWave` calls `SUBGRF_SetRfTxPower` that expects `Tx power level [0..15]` that calls `SUBGRF_SetTxParams` that expects `RF output power [-18..13] dBm`.

So it's clearly dBm, but I think there's a bug in `SUBGRF_SetRfTxPower` - it will never select `RFO_HP`

**Comment by fpistm** [comment]
> Yes, and then `RadioSetTxContinuousWave` calls `SUBGRF_SetRfTxPower` that expects `Tx power level [0..15]` that calls `SUBGRF_SetTxParams` that expects `RF output power [-18..13] dBm`.
> 
> So it's clearly dBm, but I think there's a bug in `SUBGRF_SetRfTxPower` - it will never select `RFO_HP`

Right, I guess an issue should be filled in https://github.com/STMicroelectronics/STM32CubeWL/issues

**Comment by Miceuz** [comment]
> > Yes, and then `RadioSetTxContinuousWave` calls `SUBGRF_SetRfTxPower` that expects `Tx power level [0..15]` that calls `SUBGRF_SetTxParams` that expects `RF output power [-18..13] dBm`.
> > So it's clearly dBm, but I think there's a bug in `SUBGRF_SetRfTxPower` - it will never select `RFO_HP`
> 
> Right, I guess an issue should be filled in https://github.com/STMicroelectronics/STM32CubeWL/issues

Ok, will do. So this PR then can be merged - I have altered the documentation strings to represent dBm.

**Review by fpistm — COMMENTED** [review]
LGTM. Thanks for the update.
Waiting @matthijskooijman feedback 😉

**Review by fpistm — APPROVED** [review]

**Comment by fpistm** [comment]
Thanks @Miceuz to the issue submitted here: https://github.com/STMicroelectronics/STM32CubeWL/issues/64

@matthijskooijman are you agreed with this PR now? I guess it could be merged safely as it does not impact other API.

**Review by matthijskooijman — APPROVED** [review]
I was a bit confused by all the different code snippets, but finally found the time to look a bit closer. Looks good to me now, yes.

As for `SUBGRF_SetRfTxPower()`, AFAICS the problem is not actually the code, just the documentation problem, right? Reading your comments and the code right, the function is called with dBm, passes dBm on to `SUBGRF_SetTxParams()` and the `>15` check makes sense, since LP goes up to 15dBm and HP up to 22dBm, so it should work as expected, right?

For reference, here is the upstream issue @Miceuz reported: https://github.com/STMicroelectronics/STM32CubeWL/issues/64

**Comment by fpistm** [comment]
Agreed, documentations are not coherent. Let's see how it will be handled on STM32CubeWL repo.

**Pull request merged**